internal/services/authorization/role_definition_resource.go

```diff
@@ -9,365 +9,416 @@ import (
+				}
+				stateConf := &pluginsdk.StateChangeConf{
+					Pending: []string{
+						"Pending",
+					},
+					Target: []string{
+						"OK",
+					},
+					Refresh:                   roleDefinitionUpdateStateRefreshFunc(ctx, client, id),
+					MinTimeout:                10 * time.Second,
+					ContinuousTargetOccurence: 12,
+					Timeout:                   time.Until(deadline),
+				}
+
+				if _, err := stateConf.WaitForStateContext(ctx); err != nil {
+					return fmt.Errorf("waiting for update to Role Definition %q to finish replicating", config.Name)
+				}
+			}
 
-	read, err := client.Get(ctx, scope, roleDefinitionId)
-	if err != nil {
-		return err
-	}
-	if read.ID == nil || *read.ID == "" {
-		return fmt.Errorf("Cannot read Role Definition ID for %q (Scope %q)", name, scope)
+			metadata.SetID(id)
+			return nil
+		},
 	}
-
-	d.SetId(fmt.Sprintf("%s|%s", *read.ID, scope))
-	return resourceArmRoleDefinitionRead(d, meta)
 }
 
-func resourceArmRoleDefinitionUpdate(d *pluginsdk.ResourceData, meta interface{}) error {
-	sdkClient := meta.(*clients.Client).Authorization.RoleDefinitionsClient
-	client := azuresdkhacks.NewRoleDefinitionsWorkaroundClient(sdkClient)
-	ctx, cancel := timeouts.ForUpdate(meta.(*clients.Client).StopContext, d)
-	defer cancel()
+func (r ArmRoleDefinitionResource) Read() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 5 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.Authorization.ScopedRoleDefinitionsClient
 
-	roleDefinitionId, err := parse.RoleDefinitionId(d.Id())
-	if err != nil {
-		return err
-	}
+			id, err := roledefinitions.ParseScopedRoleDefinitionID(metadata.ResourceData.Id())
+			if err != nil {
+				return err
+			}
 
-	name := d.Get("name").(string)
-	description := d.Get("description").(string)
-	roleType := "CustomRole"
-
-	permissionsRaw := d.Get("permissions").([]interface{})
-	permissions := expandRoleDefinitionPermissions(permissionsRaw)
-	assignableScopes := expandRoleDefinitionAssignableScopes(d)
-
-	properties := authorization.RoleDefinition{
-		RoleDefinitionProperties: &authorization.RoleDefinitionProperties{
-			RoleName:         utils.String(name),
-			Description:      utils.String(description),
-			RoleType:         utils.String(roleType),
-			Permissions:      &permissions,
-			AssignableScopes: &assignableScopes,
+			resp, err := client.Get(ctx, *id)
+			if err != nil {
+				if response.WasNotFound(resp.HttpResponse) {
+					return metadata.MarkAsGone(id)
+				}
+
+				return fmt.Errorf("retrieving %s: %+v", id, err)
+			}
+
+			state := RoleDefinitionModel{
+				Scope:            id.Scope,
+				RoleDefinitionId: id.RoleDefinitionId,
+			}
+
+			if model := resp.Model; model != nil {
+				// for management group, the resource id on Azure is different from the sdk generated one.
+				// so we read from the response.
+				state.RoleDefinitionResourceId = pointer.From(model.Id)
```

Comment:
Shall we parse and reforamt it?

Reply:
Is this a good reason not to do the id migration?

Reply:
I prefre to do the migration, this property was invovled in last time resource id migration..
And if we keep the original id, it will need more workaround to let the id fit typed sdk